source/particle/particle_handler.cc

```diff
@@ -222,6 +261,661 @@ namespace aspect
+
+            if (!found_cell)
+              {
+                // The particle is not in a neighbor of the old cell.
+                // Look for the new cell in the whole local domain.
+                // This case is rare.
+                try
+                  {
+                    const std::pair<const typename parallel::distributed::Triangulation<dim>::active_cell_iterator,
+                          Point<dim> > current_cell_and_position =
+                            GridTools::find_active_cell_around_point<> (*mapping,
+                                                                        *triangulation,
+                                                                        (*it)->get_location());
+                    current_cell = current_cell_and_position.first;
+                    current_reference_position = current_cell_and_position.second;
+                  }
+                catch (GridTools::ExcPointNotFound<dim> &)
+                  {
+                    // We can find no cell for this particle. It has left the
+                    // domain due to an integration error or an open boundary.
+                    continue;
+                  }
+              }
+
+            // If we are here, we found a cell and reference position for this particle
+            (*it)->set_reference_location(current_reference_position);
+
+            // Reinsert the particle into our domain if we own its cell.
+            // Mark it for MPI transfer otherwise
+            if (current_cell->is_locally_owned())
+              {
+                sorted_particles.push_back(std::make_pair(types::LevelInd(current_cell->level(),current_cell->index()),
+                                                          std::move((*it)->particle->second)));
+              }
+            else
+              {
+                const unsigned int neighbor_index = subdomain_to_neighbor_map.find(current_cell->subdomain_id())->second;
+                moved_particles[neighbor_index].push_back(*it);
+                moved_cells[neighbor_index].push_back(current_cell);
+              }
+          }
+      }
+
+      // Sort the updated particles. This pre-sort speeds up inserting
+      // them into particles to O(N) complexity.
+      std::multimap<types::LevelInd,Particle <dim,spacedim> > sorted_particles_map;
+
+      // Exchange particles between processors if we have more than one process
+      if (dealii::Utilities::MPI::n_mpi_processes(mpi_communicator) > 1)
+        send_recv_particles(moved_particles,sorted_particles_map,moved_cells);
+
+      sorted_particles_map.insert(sorted_particles.begin(),sorted_particles.end());
+
+      for (unsigned int i=0; i<particles_out_of_cell.size(); ++i)
+        remove_particle(particles_out_of_cell[i]);
+
+      particles.insert(sorted_particles_map.begin(),sorted_particles_map.end());
+    }
+
+
+
+    template <int dim, int spacedim>
+    void
+    ParticleHandler<dim,spacedim>::exchange_ghost_particles()
```

Comment:
instead of reading through this function -- could you instead use the new `parallel::GridTools::exchange_...` function Timo wrote at the hackathon for what you're doing here? that would seem simpler. Since you're not yet working with deal.II dev here, you may have to copy that function from deal.II but that shouldn't be terribly difficult.